bitbots_path_planning/bitbots_path_planning/controller.py

```diff
@@ -0,0 +1,119 @@
+        self.config_translation_slow_down_factor = self.node.declare_parameter('controller.translation_slow_down_factor', 0.5).value
+        self.config_orient_to_goal_distance = self.node.declare_parameter('controller.orient_to_goal_distance', 1.0).value
+
+    def _get_yaw(self, pose) -> float:
+        return euler_from_quaternion(numpify(pose))[2]
+
+    def step(self, path: Path) -> Twist:
+        # Starting point for our goal velocity calculation
+        cmd_vel = Twist()
+
+        # Create an default pose in the origin of our base footprint
+        pose_geometry_msgs = PoseStamped()
+        pose_geometry_msgs.header.frame_id = self.node.get_parameter("base_footprint_frame").value
+
+        # We get our pose in respect to the map frame (also frame of the path message)
+        # by transforming the pose above into this frame
+        try:
+            current_pose: Pose = self.buffer.transform(pose_geometry_msgs, path.header.frame_id).pose
+        except (tf2.ConnectivityException, tf2.LookupException, tf2.ExtrapolationException) as e:
+                self.node.get_logger().warn('Failed to perform controller step: ' + str(e))
+                return cmd_vel
+
+        # Select pose for the carrot (goal_pose for this controller) and the end pose of the global plan
+        # End conditions with an empty or shorter than carrot distance path need to be considered
+        if len(path.poses) > 0:
+            end_pose: Pose = path.poses[-1].pose
+            if len(path.poses) > self.config_carrot_distance:
+                goal_pose: Pose = path.poses[self.config_carrot_distance].pose
+            else:
+                goal_pose: Pose = end_pose
+        else:
+            return cmd_vel
+
+        # Calculate the intended walk angle based on the carrot and our position
+        walk_angle = math.atan2(
+            goal_pose.position.y - current_pose.position.y,
+            goal_pose.position.x - current_pose.position.x)
+
+        # Calculate the heading angle from our current position to the final position of the global plan
+        final_walk_angle = math.atan2(
+            end_pose.position.y - current_pose.position.y,
+            end_pose.position.x - current_pose.position.x)
+
+        # Calculate the distance from our current position to the final position of the global plan
+        distance = math.hypot(
+            end_pose.position.x - current_pose.position.x,
+            end_pose.position.y - current_pose.position.y)
+
+        # Calculate the translational walk velocity. It considers the distance and breaks if we are close to the final position of the global plan
+        walk_vel = min(distance * self.config_translation_slow_down_factor, self.config_max_vel_x)
+
+        # Check if we are so close to the final position of the global plan that we want to align us with its orientation and not the heading towards its position
```

Comment:
Instead of `heading towards the final position`, I would prefer `heading on the path towards the final position`.

Reply:
Is this the correct line?